File: ios/SalonBoothMath/FreeAccessState.swift
import Foundation

enum PremiumFeature {
    case save
    case history
    case compare
    case payCheckup
}

@MainActor
final class FreeAccessState: ObservableObject {
    @Published private(set) var usedSave: Bool
    @Published private(set) var usedHistory: Bool
    @Published private(set) var usedCompare: Bool
    @Published private(set) var usedPayCheckup: Bool

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        usedSave = defaults.bool(forKey: Keys.save)
        usedHistory = defaults.bool(forKey: Keys.history)
        usedCompare = defaults.bool(forKey: Keys.compare)
        usedPayCheckup = defaults.bool(forKey: Keys.payCheckup)
    }

    func used(_ feature: PremiumFeature) -> Bool {
        switch feature {
        case .save: return usedSave
        case .history: return usedHistory
        case .compare: return usedCompare
        case .payCheckup: return usedPayCheckup
        }
    }

    func claim(_ feature: PremiumFeature, isUnlocked: Bool) -> Bool {
        if isUnlocked { return true }
        if used(feature) { return false }
        setUsed(feature)
        return true
    }

    private func setUsed(_ feature: PremiumFeature) {
        switch feature {
        case .save:
            usedSave = true
            defaults.set(true, forKey: Keys.save)
        case .history:
            usedHistory = true
            defaults.set(true, forKey: Keys.history)
        case .compare:
            usedCompare = true
            defaults.set(true, forKey: Keys.compare)
        case .payCheckup:
            usedPayCheckup = true
            defaults.set(true, forKey: Keys.payCheckup)
        }
    }

    private enum Keys {
        static let save = "didUseFreeSave"
        static let history = "didUseFreeHistory"
        static let compare = "didUseFreeCompare"
        static let payCheckup = "didUseFreePayCheckup"
    }
}

File: android/app/src/main/java/com/everittventures/salonboothmath/FreeAccessState.kt
package com.everittventures.salonboothmath

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue

enum class PremiumFeature {
    SAVE,
    HISTORY,
    COMPARE,
    PAY_CHECKUP
}

class FreeAccessState(context: Context) {
    private val prefs = context.getSharedPreferences("salon_booth_math", Context.MODE_PRIVATE)

    private var usedSave by mutableStateOf(prefs.getBoolean(KEY_SAVE, false))
    private var usedHistory by mutableStateOf(prefs.getBoolean(KEY_HISTORY, false))
    private var usedCompare by mutableStateOf(prefs.getBoolean(KEY_COMPARE, false))
    private var usedPayCheckup by mutableStateOf(prefs.getBoolean(KEY_PAY_CHECKUP, false))

    fun used(feature: PremiumFeature): Boolean = when (feature) {
        PremiumFeature.SAVE -> usedSave
        PremiumFeature.HISTORY -> usedHistory
        PremiumFeature.COMPARE -> usedCompare
        PremiumFeature.PAY_CHECKUP -> usedPayCheckup
    }

    fun claim(feature: PremiumFeature, isUnlocked: Boolean): Boolean {
        if (isUnlocked) return true
        if (used(feature)) return false
        setUsed(feature)
        return true
    }

    private fun setUsed(feature: PremiumFeature) {
        val key = when (feature) {
            PremiumFeature.SAVE -> {
                usedSave = true
                KEY_SAVE
            }
            PremiumFeature.HISTORY -> {
                usedHistory = true
                KEY_HISTORY
            }
            PremiumFeature.COMPARE -> {
                usedCompare = true
                KEY_COMPARE
            }
            PremiumFeature.PAY_CHECKUP -> {
                usedPayCheckup = true
                KEY_PAY_CHECKUP
            }
        }
        prefs.edit().putBoolean(key, true).apply()
    }

    private companion object {
        const val KEY_SAVE = "did_use_free_save"
        const val KEY_HISTORY = "did_use_free_history"
        const val KEY_COMPARE = "did_use_free_compare"
        const val KEY_PAY_CHECKUP = "did_use_free_pay_checkup"
    }
}

@Composable
fun rememberFreeAccessState(context: Context): FreeAccessState =
    remember(context) { FreeAccessState(context) }
